crates/uv-resolver/src/lock.rs

```diff
@@ -716,6 +756,80 @@ impl SourceDist {
     }
 }
 
+impl From<GitReference> for GitSourceKind {
+    fn from(value: GitReference) -> Self {
+        match value {
+            GitReference::Branch(branch) => GitSourceKind::Branch(branch.to_string()),
+            GitReference::Tag(tag) => GitSourceKind::Tag(tag.to_string()),
+            GitReference::ShortCommit(rev) => GitSourceKind::Rev(rev.to_string()),
+            GitReference::BranchOrTag(rev) => GitSourceKind::Rev(rev.to_string()),
+            GitReference::BranchOrTagOrCommit(rev) => GitSourceKind::Rev(rev.to_string()),
+            GitReference::NamedRef(rev) => GitSourceKind::Rev(rev.to_string()),
+            GitReference::FullCommit(rev) => GitSourceKind::Rev(rev.to_string()),
+            GitReference::DefaultBranch => GitSourceKind::DefaultBranch,
+        }
+    }
+}
+
+impl From<GitSourceKind> for GitReference {
+    fn from(value: GitSourceKind) -> Self {
+        match value {
+            GitSourceKind::Branch(branch) => GitReference::Branch(branch),
+            GitSourceKind::Tag(tag) => GitReference::Tag(tag),
+            GitSourceKind::Rev(rev) => GitReference::from_rev(&rev),
+            GitSourceKind::DefaultBranch => GitReference::DefaultBranch,
+        }
+    }
+}
+
+/// Construct the lockfile-compatible [`URL`] for a [`GitSourceDist`].
+fn locked_git_url(git_dist: &GitSourceDist) -> Url {
+    let mut url = git_dist.git.repository().clone();
+
+    // Clear out any existing state.
+    url.set_fragment(None);
+    url.set_query(None);
+
+    // Put the subdirectory in the query.
+    if let Some(subdirectory) = git_dist.subdirectory.as_deref().and_then(Path::to_str) {
+        url.query_pairs_mut()
+            .append_pair("subdirectory", subdirectory);
+    }
+
+    // Put the requested reference in the query.
+    match git_dist.git.reference() {
+        GitReference::Branch(branch) => {
+            url.query_pairs_mut()
+                .append_pair("branch", branch.to_string().as_str());
+        }
+        GitReference::Tag(tag) => {
+            url.query_pairs_mut()
+                .append_pair("tag", tag.to_string().as_str());
+        }
+        GitReference::ShortCommit(rev)
+        | GitReference::BranchOrTag(rev)
+        | GitReference::BranchOrTagOrCommit(rev)
+        | GitReference::NamedRef(rev)
+        | GitReference::FullCommit(rev) => {
+            url.query_pairs_mut()
+                .append_pair("rev", rev.to_string().as_str());
+        }
+        GitReference::DefaultBranch => {}
+    }
+
+    // Put the precise commit in the fragment.
+    url.set_fragment(
+        git_dist
+            .git
+            .precise()
+            .as_ref()
+            .map(GitSha::to_string)
+            .as_deref(),
+    );
```

Comment:
I think it should be an error if we don't have a precise commit right? Without the precise commit revision, the lock file cannot guarantee reproducibility.